test/cpp/end2end/xds_end2end_test.cc

```diff
@@ -475,109 +477,69 @@ class AdsServiceImpl : public AggregatedDiscoveryService::Service,
-        &subscriber_state);
-  }
-
-  // Resource unsubscription:
-  // 1. update the entry in the resources map indexed
-  // by resource type/name pair to remove this subscription
-  // 2. remove this entry from the subscription map.
-  // 3. remove this resource type from the subscription map if there are no more
-  // resources subscribed for the resource type.
-  void ResourceUnsubscribe(const std::string& resource_type,
-                           const std::string& name,
-                           SubscriptionMap* subscription_map) {
-    auto subscription_by_type_it = subscription_map->find(resource_type);
-    if (subscription_by_type_it == subscription_map->end()) {
-      gpr_log(GPR_INFO, "ADS[%p]: resource type %s not subscribed", this,
-              resource_type.c_str());
-      return;
-    }
-    auto& subscription_by_type_map = subscription_by_type_it->second;
-    auto subscription_it = subscription_by_type_map.find(name);
-    if (subscription_it == subscription_by_type_map.end()) {
-      gpr_log(GPR_INFO, "ADS[%p]: resource name %s of type %s not subscribed",
-              this, name.c_str(), resource_type.c_str());
-      return;
-    }
-    gpr_log(GPR_INFO,
-            "ADS[%p]: Unsubscribe to resource type %s name %s state %p", this,
-            resource_type.c_str(), name.c_str(), &subscription_it->second);
-    auto resource_by_type_it = resources_map_.find(resource_type);
-    GPR_ASSERT(resource_by_type_it != resources_map_.end());
-    auto& resource_by_type_map = resource_by_type_it->second;
-    auto resource_it = resource_by_type_map.find(name);
-    GPR_ASSERT(resource_it != resource_by_type_map.end());
-    resource_it->second.subscribers.erase(&subscription_it->second);
-    if (resource_it->second.subscribers.empty() &&
-        !resource_it->second.resource.has_value()) {
-      gpr_log(GPR_INFO,
-              "ADS[%p]: Erasing resource type %s name %s from resource map "
-              "since there are no more subscribers for this unset resource",
-              this, resource_type.c_str(), name.c_str());
-      resource_by_type_map.erase(resource_it);
-    }
-    subscription_by_type_map.erase(subscription_it);
-    if (subscription_by_type_map.empty()) {
-      gpr_log(GPR_INFO,
-              "ADS[%p]: Erasing resource type %s from subscription_map", this,
-              resource_type.c_str());
-      subscription_map->erase(subscription_by_type_it);
+  // Subscribes to a resource if not already subscribed:
+  // 1. Sets the update_queue field in subscription_state.
+  // 2. Adds subscription_state to resource_state->subscriptions.
+  void MaybeSubscribe(const std::string& resource_type,
+                      const std::string& resource_name,
+                      SubscriptionState* subscription_state,
+                      ResourceState* resource_state,
+                      UpdateQueue* update_queue) {
+    if (subscription_state->update_queue != nullptr) return;
```

Comment:
can we add a comment or a log to remind us that update_queue already set is an indication we have already subscribed to the resource.

or we can add the check before calling this method

Reply:
I'll add the comment in a subsequent PR so that I don't have to wait for the tests to run again.